paddle/framework/op_kernel_type.h

```diff
@@ -85,9 +85,14 @@ inline std::string KernelTypeToString(const OpKernelType& kernel_key) {
   return stream.str();
 }
 
+inline bool NeedTransformLayout(const DataLayout& l, const DataLayout& r) {
+  return l != DataLayout::kAnyLayout && r != DataLayout::kAnyLayout && l != r;
+}
+
 inline bool TransFromNeeded(const OpKernelType& l, const OpKernelType& r) {
   return (!platform::places_are_same_class(l.place_, r.place_)) ||
-         (l.data_type_ != r.data_type_) || (l.data_layout_ != r.data_layout_);
+         (l.data_type_ != r.data_type_) ||
```

Comment:
**I believe we are in the wrong way.** 

let's start from a simple convolution case:
data -> conv2d -> batch_norm -> fc -> softmax
we may have a quite different layout pipelines choices.
NHWC->NCHW -> NCHW -> NCHW
NHWC->NHWC -> NCHW -> NCHW
NHWC->NHWC -> NHWC -> NCHW
....
NCHW->NCHW -> NCHW -> NCHW
NCHW->NHWC -> NCHW -> NCHW
....
All of them are legal, but only one of them is the best.  in cudnn, NHWC is 20% faster than other formats.
So, the answer to select one from above, it's would be NHWC pipeline, no matter what the data format is, just transform the data into NHWC, then go through the data pipeline.

In the case above, I want to claim that **we do not run on mixed devices, we can run on different device type.**.  
It means the **data pipeline layout** can be determined once we have the device information, we only need to transform data before or after the data pipeline. 
data pipeline refer to `conv2d -> batch_norm -> fc -> softmax`, without data input and output.

Reply:
To be clear, we support multiple device, but not mixed device.

Reply:
yes, I Agree with you, we have a data process chain, and there should be some best practice. On the other hand, I also think that the current data transform just does transform according to expected_kernel_type and actual_kernel_type_for_var, it does not consider what kind of expected_kernel_type it should use, I think that should be implemented in GetExpectedKernelType by the user or by framework. So it does not conflict with the current implementation.